### ISSUE TITLE
silly_div : I can't understand why b is 33. I'm silly

### ISSUE THREAD

**Issue by gnahznux** [hellerve-project-based-learning/silly#1]
I wonder why b is 33 instead of 32.
```
silly silly_div(silly x, silly y) {
  // disregard the sign
  uint64_t sign_bit = 1UL<<63;
  uint64_t rem = x & ~sign_bit;
  uint64_t div = y & ~sign_bit;

  uint64_t quo = 0UL;
  int b = 33; // 64 / 2 + 1

  while (rem && b >= 0) {
      uint64_t d = rem / div;
      rem %= div;
      quo += d << b;

      rem <<= 1;
      --b;
  }

  uint64_t res = quo >> 1;
  // add the sign
  res |= (x & sign_bit) ^ (y & sign_bit);

  return res;
}
```

**Comment by hellerve** [comment]
`b` is the number of bits we go through. The upper bound for this (unless `rem` shortcuts us) is `<width> / 2 + 1`, where `width` is the width of the number. Since we know that that width is always 64 bits, we can just hardcode it in our function.

I encourage you to [read through my blog post about this](https://blog.veitheller.de/Fixed_Point_Division.html) to get a better sense as to why.

**Comment by gnahznux** [comment]
thanks for your reply !!!
I have read your blog before asking the question. It's very impressive.
That **the upper bound is 64/2 +1**  means we go through 33 bit at the most? Can you explain specific?
And after while loop , ```quo >> 1``` is weird . 
If  b is 32 , we don't need >> 1 after while loop, am I right?

**Comment by hellerve** [comment]
I haven’t thought about this code in quite some time, but you seem to be correct. IIRC the whole thing has something to do with rounding (also the increment of `quo` that I commented out in the code). I can’t replicate the behavior we’re fixing here, however, so the only answer that I have now is: I forgot why I did what I did there (I realize  that this is a disappointment, but such is life with code that you wrote a few years ago).

**Comment by gnahznux** [comment]
Don't feel upset about it (doge). I will tell you the answer when I figure it out.

**Comment by nilpunch** [comment]
In my tests, this number **must** be equal _fractionalPlaces + 1_.
For 32-bit fractional part, that's 33.
For 20-bit fractional part, that's 21.

Algorithm gives wrong results otherwise. Just divide by 1 anything to see the issue.

**Comment by hellerve** [comment]
I’ll take a look. Given that it’s a toy library I’ve written 8 years ago now, it might take me a while to get it :) Thanks for the input!

**Comment by hellerve** [comment]
I’m not quite able to reproduce what you are saying. If I set the number of bits to traverse to 32 and remove `quo >> 1` from the algorithm, the results seem to be correct, even if divide anything by 1 or 1 by anything (wasn’t quite sure which one you meant). What am I missing?

**Comment by nilpunch** [comment]
In my library, I have a variable number of fractional places controlled by a constant.
Try experimenting with this algorithm using different combinations of integer and fractional places to reproduce the results.
BTW, I'm in search for a better algorithm, but this one works for now. :)

**Comment by hellerve** [comment]
I did a little bit of thinking about this. If we used `32`, we wouldn’t have the rounding precision we want. So basically we need 32 bits of precision and one bit to round accurately. We could do without it, but then the right-hand side would lose precision on rounding.

Consider, for instance, 1/7. We’d have a quotient that, at position 33, has the bit set. Without accurate rounding, we’d get `0.14285714272409678`. With rounding, we get `0.14285714295692742`.

Having more bits does not add anything, by the way, because for rounding that doesn’t matter anymore.

Does that make sense? If so, I’d close this issue.

**Comment by hellerve** [comment]
> Consider, for instance, 1/7. We’d have a quotient that, at position 33, has the bit set. Without accurate rounding, we’d get `0.14285714272409678`. With rounding, we get `0.14285714295692742`.

I got concerned about the veracity of my numbers here, so I asked ChatGPT to verify them. It produced the following script:

```python
from decimal import Decimal, getcontext

# Set high precision to simulate fixed point with many bits
getcontext().prec = 50

def check_fixed_point_division():
    # Simulate 32-bit fractional fixed point
    fractional_bits = 32

    # Value to divide: 1 / 7
    x = Decimal(1)
    y = Decimal(7)

    # Naive fixed-point encoding of 1
    x_fixed = int(x * (1 << fractional_bits))
    y_fixed = int(y * (1 << fractional_bits))

    # Division in fixed point with 32-bit shift
    result_32 = (x_fixed << fractional_bits) // y_fixed
    result_32_float = Decimal(result_32) / (1 << fractional_bits)

    # Division in fixed point with 33-bit shift (for rounding)
    result_33 = (x_fixed << (fractional_bits + 1)) // y_fixed
    rounded_result_33 = (result_33 + 1) >> 1
    result_rounded_float = Decimal(rounded_result_33) / (1 << fractional_bits)

    # Check the hex representation too
    hex_32 = hex(result_32)
    hex_rounded = hex(rounded_result_33)

    return {
        "32-bit result (hex)": hex_32,
        "32-bit result (float)": float(result_32_float),
        "33-bit rounded result (hex)": hex_rounded,
        "33-bit rounded result (float)": float(result_rounded_float),
    }

check_fixed_point_division()
```

Use at your own peril, this is LLM-generated.

**Comment by nilpunch** [comment]
> Does that make sense? If so, I’d close this issue.

Yeah, my thoughts were actually the same, but for some reason, my test cases were failing when I hardcoded the value as `64 / 2 + 1` and started tweaking the number of fractional bits in the fixed-point representation.

And yes, you can definitely close the issue. I ended up making a faster algorithm, so I kind of abandoned the old one.

Here it is, in case you're interested.
It's based on the naive formula `result = (x << FractionalBits) / y`, but tweaked to avoid wrong results with large `x` values. It uses the LZC instruction to count how many bits `x` can be shifted left without losing significant bits, and applies the remaining shift to `y` (to the right), dropping some insignificant bits from the denominator. With this method, precision loss will occur only when the numerator and denominator have significantly different magnitudes (`x` much bigger than `y`).

Looks something like:
```
x_lzc = LeadingZeroCount(x)

x_shift = x_lzc > fractional_bits ? fractional_bits : x_lzc 
y_shift = fractional_bits - x_shift

x_scaled = x << x_shift
y_scaled = y >> y_shift

result = x_scaled / y_scaled
```

It misses some sign fiddling and overflow checking, but adding that is basic, which I did [here](https://github.com/nilpunch/fixed-point/blob/8380493e3074dea49b2087420412c3115a70a972/Runtime/FP.Raw.cs#L107).
